### PR TITLE
feat(api-idorslug) Finished testing all `convert_args` with Resouce Ids & Slugs

### PR DESCRIPTION
**Pull request by iamrajjoshi**
* Finished testing the rest of the `convert_args` ❗  
* Modified `extract_all_url_patterns` to handle edge-case where `sentry-api-0-organization-monitor-check-in-attachment` calls `method_dispatch` to use a different class for each method.
* Added the  a property to `method_dispatch` that `extract_all_url_patterns` could user
* Updated some types

For: https://github.com/getsentry/team-core-product-foundations/issues/194

### CODE REVIEW

**Review by sentaur-athena — APPROVED** [review]

File: src/sentry/monitors/endpoints/monitor_ingest_checkin_attachment.py [hunk 1]
```diff
@@ -132,7 +137,13 @@ def convert_args(
...
         # When looking up via GUID we do not check the organization slug,
         # validate that the slug matches the org of the monitors project
-        if organization_slug and project.organization.slug != organization_slug:
+        if not organization_slug:
+            raise ResourceDoesNotExist
+
+        if options.get("api.id-or-slug-enabled") and str(organization_slug).isnumeric():
+            if project.organization.id != organization.id:
+                raise ResourceDoesNotExist
+        elif project.organization.slug != organization_slug:
```
Comment (schew2381) [inline src/sentry/monitors/endpoints/monitor_ingest_checkin_attachment.py:146]: I think this logic is a bit different than what was originally there. We only raise if the `organization_slug` was set and it doesn't match.

This is a bit implicit b/c we're not checking the option, but rather relying on the fact that both slug and id are unique. Prob good to add a comment explaining this?
```suggestion
        if organization_slug and project.organization.slug != organization_slug and project.organization.id != organization_slug:
                raise ResourceDoesNotExist
```

File: tests/sentry/api/path_params/resources/test_project_slug.py [hunk 2]
```diff
@@ -77,3 +80,93 @@ def project_codeowners_details_test(self, endpoint_class, slug_params, *args):
...
+            trigger_type=0,
+            value=1,
+            window_in_seconds=1,
+            project=self.project,
+        )
+
+        non_slug_mappings: dict[str, Any] = {
+            "release_threshold": release_threshold.id,
+        }
+
```
Comment (schew2381) [inline tests/sentry/api/path_params/resources/test_project_slug.py:101]:
```suggestion
```

File: tests/sentry/api/path_params/resources/test_project_slug.py [hunk 3]
```diff
@@ -77,3 +80,93 @@ def project_codeowners_details_test(self, endpoint_class, slug_params, *args):
...
+            project=self.project,
+            label="test",
+            data={},
+            environment_id=None,
+        )
+
+        non_slug_mappings: dict[str, Any] = {
+            "rule_id": str(rule.id),
+        }
+
```
Comment (schew2381) [inline tests/sentry/api/path_params/resources/test_project_slug.py:133]:
```suggestion
```

File: src/sentry/monitors/endpoints/monitor_ingest_checkin_attachment.py (hunk 1 above)
Comment (iamrajjoshi) [reply src/sentry/monitors/endpoints/monitor_ingest_checkin_attachment.py:146]: good catch! 🥔

File: src/sentry/api/utils.py [hunk 4]
```diff
@@ -354,6 +354,10 @@ def dispatcher(request, *args, **kwargs):
         handler = dispatch_mapping.get(request.method, invalid_method)
         return handler(request, *args, **kwargs)
 
+    # This allows us to surface the mapping when iterating through the URL patterns
+    # Check `tes_id_or_slug_path_params.py` for usage
```
Comment (schew2381) [inline src/sentry/api/utils.py:358]:
```suggestion
    # Check `test_id_or_slug_path_params.py` for usage
```

File: src/sentry/monitors/endpoints/monitor_ingest_checkin_attachment.py [hunk 5]
```diff
@@ -132,7 +137,20 @@ def convert_args(
...
+            organization_slug
+            and project.organization.slug != organization_slug
+            and project.organization.id != organization_slug
+        ):
+            raise ResourceDoesNotExist
+
+        if options.get("api.id-or-slug-enabled") and str(organization_slug).isnumeric():
+            if project.organization.id != organization.id:
+                raise ResourceDoesNotExist
+        elif project.organization.slug != organization_slug:
```
Comment (schew2381) [inline src/sentry/monitors/endpoints/monitor_ingest_checkin_attachment.py:153]: I think we can remove this part

File: tests/sentry/api/path_params/test_id_or_slug_path_params.py [hunk 6]
```diff
@@ -206,15 +279,8 @@ def test_if_endpoints_work_with_id_or_slug(self):
...
-                    ):
-                        self.convert_args_setup_registry[callback.convert_args](
-                            callback, slug_path_params
-                        )
-                    # Temporary for "project_slug" endpoints
-                    elif convert_args := self.convert_args_setup_registry.get(
-                        callback.convert_args
-                    ):
+                        convert_args := self.convert_args_setup_registry.get(callback.convert_args)
+                    ) is not None:
```
Comment (schew2381) [inline tests/sentry/api/path_params/test_id_or_slug_path_params.py:283]: nit: 
```suggestion
                        convert_args := self.convert_args_setup_registry.get(callback.convert_args)
                    ):
```

**Review by schew2381 — APPROVED** [review]
lgtm overall, some small nits

**Pull request merged**